src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,885 @@
+/// <reference path="../refactorProvider.ts" />
+/// <reference path="../../compiler/checker.ts" />
+
+/* @internal */
+namespace ts.refactor.extractMethod {
+    export const name = "extract_method";
+    const extractMethod: Refactor = {
+        name: "Extract Method",
+        description: Diagnostics.Extract_function.message,
+        getAvailableActions,
+        getEditsForAction,
+    };
+
+    registerRefactor(extractMethod);
+
+    /** Compute the associated code actions */
+    function getAvailableActions(context: RefactorContext): ApplicableRefactorInfo[] | undefined {
+        const rangeToExtract = getRangeToExtract(context.file, { start: context.startPosition, length: context.endPosition - context.startPosition });
+
+        const targetRange: TargetRange = rangeToExtract.targetRange;
+        if (targetRange === undefined) {
+            return undefined;
+        }
+
+        const extractions = extractRange(targetRange, context.file, context);
+        if (extractions.length === 0) {

```

Comment:
It seems a little strange that `extractRange` returns empty for "failure" and then this method, which also returns an array, converts it into `undefined`.